docs/proposals/20190704-AutomatedUpgrades.md

```diff
@@ -0,0 +1,196 @@
+In the case of cloud installs, or where users may want to boot without PXE/iPXE, we will require a boot partition.
+In all cases we require a data partition.
+
+I also propose we rename the `DATA` partition to `EPHEMERAL` to make it clear that this partition should not be used to store application data.
+Our suggestion to users who do want persistent storage on the host is to mount another disk and to use that.
+Talos will make the guarantee to only manage the "owned" block device.
+
+### Performing Upgrades in a Container
+
+The move to running `rootfs` in memory, and thus making the block device ephemeral, and the split of `init` into `init` and `machined`, allows us to take an approach to upgrades that is clean, and relatively low risk.
+The workflow will look roughly like the following:
+
+- an upgrade request is received by `osd` and proxied to `machined`
+- `machined` stops `system` and `k8s.io` services
+- `machined` unmounts all mounts in `/var`
+- `machined` runs the specified installer container
+
+### Automated Upgrades
+
+All of the aformentioned improvements help in the higher goal of automated upgrades.
+Once the above is implemented, upgrades should be simple and self-contained to a machine.
+By running upgrades in containers, we can have very simple operator implementation that can rollout an upgrade across a cluster by simply:
+
+- checking for a new tag
+- orchestrating cluster-wide rollout
+
+The operator will initially coordinate upgrades only, but it will also provide a basis for a number of future automation opportunities.
+
+#### Release Channels
+
+We will maintain a service that allows for the operator to discover when new releases of Talos are made available.
+There will be three "channels" that a machine can subscribe to:
+
+- alpha
+- beta
+- stable
+
+The implementation details around this service and its' interaction with the operator is not in the scope of this proposal.
+
+#### Workflow
+
+In order to upgrade nodes, the operator will subscribe to events for nodes.
+It will keep an in-memory list of all nodes.
+
+The operator will periodically query the target registry for a list of containers within a time range.
+The time range should start where a previous query has ended.
+The initial query should use the currently installed image's timestamp to decide where to start.
+
+In order to supply enough information to the operator, such that it can make an informed decision about using a particular container, we will make use of labels.
+With labels we can communicate to the operator that a container is:
+
+1. an official release
+2. within a given channel
+3. contains a version that is compatible with the current cluster
+
+There may be more information we end up needing, the above is just an example.
+
+Once a container has been chosen, the operator will then create a schedule based on polices.
+The initial policies will perform an upgrade based on:
+
+1. x% of nodes at a given time
```

Comment:
I'd recommend using discrete node counts rather than percentages.  In general, those are better understood and more in-line with dimensioning for all but the largest of clusters.

Reply:
Good point. I think I might have this covered a few bullet points down. There is an extra time element in my proposal but we could easily have a case where users can specify a time of 0, indicating that a rollout should happen one after the other.